kernel/Base/MeshEntity.h

```diff
@@ -0,0 +1,145 @@
+/*
+ This file forms part of hpGEM. This package has been developed over a number of
+ years by various people at the University of Twente and a full list of
+ contributors can be found at http://hpgem.org/about-the-code/team
+
+ This code is distributed using BSD 3-Clause License. A copy of which can found
+ below.
+
+
+ Copyright (c) 2021, University of Twente
+ All rights reserved.
+
+ Redistribution and use in source and binary forms, with or without
+ modification, are permitted provided that the following conditions are met:
+
+ 1. Redistributions of source code must retain the above copyright notice, this
+ list of conditions and the following disclaimer.
+
+ 2. Redistributions in binary form must reproduce the above copyright notice,
+ this list of conditions and the following disclaimer in the documentation
+ and/or other materials provided with the distribution.
+
+ 3. Neither the name of the copyright holder nor the names of its contributors
+ may be used to endorse or promote products derived from this software without
+ specific prior written permission.
+
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef HPGEM_MESHENTITY_H
+#define HPGEM_MESHENTITY_H
+
+#include "functional"
+#include <type_traits>
+
+namespace hpgem {
+namespace Base {
+
+// Predeclaration to prevent circular dependency issues
+class Edge;
+class Element;
+class Face;
+class Node;
+
+/**
+ * Visitor for the MeshEntity
+ * @tparam constify Whether the MeshEntities are const or not
+ */
+template <bool constify = false>
+class MeshEntityVisitor {
+   public:
+    virtual ~MeshEntityVisitor() = default;
+    /// Helper to add const
+    template <typename T>
+    using Constified =
+        typename std::conditional<constify, typename std::add_const<T>::type,
+                                  T>::type;
+
+    virtual void visit(Constified<Base::Element>& element) = 0;
+    virtual void visit(Constified<Base::Edge>& edge) = 0;
+    virtual void visit(Constified<Base::Face>& face) = 0;
+    virtual void visit(Constified<Base::Node>& node) = 0;
+};
+
+using ConstMeshEntityVisitor = MeshEntityVisitor<true>;
```

Comment:
I actually would either typedef both cases or use C++17 where you do not need the `<>`anymore